cclib/method/ddec.py

```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
+        # Assign rho^cond on grid using generator object
+        self.rho_cond = copy.deepcopy(self.chgdensity)
+        self.rho_cond.data = numpy.zeros_like(self.rho_cond.data, dtype=float)
+        rho_cond_sqrt = numpy.zeros_like(self.rho_cond.data, dtype=float)
+
+        # Generator object to iterate over the grid
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        atoms = len(self.data.atomnos)
+        indices = (
+            (i, x, y, z)
+            for i in range(atoms)
+            for x in range(xshape)
+            for y in range(yshape)
+            for z in range(zshape)
+        )
+
+        self._leftterm = numpy.zeros((atoms, xshape, yshape, zshape), dtype=float)
+        self.tau = []
+
+        # rho_cond -- equation 65 in doi: 10.1039/c6ra04656h
+        for atomi, xindex, yindex, zindex in indices:
+            self.rho_cond.data[xindex][yindex][zindex] = (
+                self.rho_cond.data[xindex][yindex][zindex]
+                + self._cond_density[atomi][self.closest_r_index[atomi][xindex][yindex][zindex]]
+            )
+
+        rho_cond_sqrt = numpy.sqrt(self.rho_cond.data)
+
+        for atomi in range(len(self.data.atomnos)):
+            self.tau.append(numpy.zeros_like(self.proatom_density[atomi], dtype=float))
+            grid = ((x, y, z) for x in range(xshape) for y in range(yshape) for z in range(zshape))
+            for xindex, yindex, zindex in grid:
+                # leftterm is the first spherical average term in equation 66.
+                # <rho^cond_A(r_A) / sqrt(rho^cond(r))>
+                self._leftterm[atomi][xindex][yindex][zindex] = self._cond_density[atomi][
+                    self.closest_r_index[atomi][xindex][yindex][zindex]
+                ]
+            self._leftterm[atomi] = self._leftterm[atomi] / rho_cond_sqrt
+            for radiusi in range(len(self.tau[atomi])):
+                grid_filter = self.closest_r_index[atomi] == radiusi
+                num_grid_filter = numpy.count_nonzero(grid_filter)
+                if num_grid_filter < 1:
+                    self.tau[atomi][radiusi] = 0.0
+                else:
+                    leftaverage = numpy.sum(grid_filter * self._leftterm[atomi]) / num_grid_filter
+                    rightaverage = numpy.sum(grid_filter * rho_cond_sqrt) / num_grid_filter
+                    if leftaverage < 1e-20:
+                        self.tau[atomi][radiusi] = 0.0
+                    else:
+                        self.tau[atomi][radiusi] = numpy.divide(
+                            leftaverage,
+                            rightaverage,
+                            out=numpy.zeros_like(leftaverage),
+                            where=rightaverage != 0.0,
+                        )
+            # Make tau monotonic decreasing
+            self.tau[atomi] = numpy.maximum.accumulate(self.tau[atomi][::-1])[::-1]
+
+    def _psiai(self, ya, bigpsiAI, atomi):
```

Comment:
Docstring for where this comes from (second box in S1?)

Reply:
Oh, it's S100, better to say that.

Reply:
I've added this. Thanks!